src/bytes_mut.rs

```diff
@@ -1266,6 +1288,38 @@ fn original_capacity_from_repr(repr: usize) -> usize {
     1 << (repr + (MIN_ORIGINAL_CAPACITY_WIDTH - 1))
 }
 
+#[inline]
+fn check_capacity_overflow(capacity: usize) {
+    if mem::size_of::<usize>() < 8 && capacity > isize::MAX as usize {
+        capacity_overflow()
+    }
+}
+
+fn capacity_overflow() -> ! {
+    panic!("capacity overflow")
+}
+
+#[inline]
+fn alloc_buf(capacity: usize) -> NonNull<u8> {
+    if capacity == 0 {
+        NonNull::dangling()
+    } else {
+        check_capacity_overflow(capacity);
+        unsafe {
```

Comment:
If this function (`alloc_buf`) is going to be safe to call, it might be worth having a comment here explaining why this is always safe with any value of `capacity`.